product_contract/models/contract_line.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2017 LasLabs Inc.
+# Copyright 2018 ACSONE SA/NV.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo import api, fields, models
+
+
+class ContractLine(models.Model):
+    _inherit = "contract.line"
+    _rec_name = "display_name"
+
+    sale_order_line_id = fields.Many2one(
+        comodel_name="sale.order.line",
+        string="Sale Order Line",
+        required=False,
+        copy=False,
+    )
+
+    def _prepare_invoice_line(self):
+        res = super(ContractLine, self)._prepare_invoice_line()
+        if self.sale_order_line_id and res:
+            res["sale_line_ids"] = [(6, 0, [self.sale_order_line_id.id])]
```

Comment:
FYI, there is now a Command in odoo.fields in order to avoid magic numbers and improve readability